File: shapes.py
import abc
import numpy as np
from itertools import product, ifilter, izip

from quaternion import Quaternion, Q

def sameside3d(p, ref, A,B,C ):
    cpAB_BC = np.cross(A - B, A - C)

    sameABCr = np.dot(ref - A, cpAB_BC)
    sameABCp = np.dot(  p - A, cpAB_BC)

    sameABC = any(( sameABCr < 0 and sameABCp < 0 
                  , sameABCr > 0 and sameABCp > 0 
                  ))

    if sameABC:
        return True

    return False


class Rotateable():
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def points(self): 
        " Return constructible points in a list. "
        pass

    @abc.abstractmethod
    def set_points(self, points):
        " Construct shape from list of points. "
        pass

    def rotate3D_qua(self, r_angles, origin=np.array([0.0,0.0,0.0])):
        " Rotate the shape by [x,y,z] degrees about origin of [x,y,x], using quaternions"
        r_rads = r_angles * np.pi / np.float(180.0)

        axis_unit = ( np.cos(r_angles[0])
                    , np.sin(r_angles[0]) * np.cos(r_angles[1])
                    , np.sin(r_angles[0]) * np.sin(r_angles[1])
                    )

        rot_q = Q.from_axisangle(r_rads[2], axis_unit)

        new_points = [ (origin + (rot_q * (point - origin)))
                       for point in self.points() 
                     ]
        self.set_points(new_points)
        return self
            

    def rotate3D(self, r_angles, origin=np.array([0.0,0.0,0.0])):
        " Rotate the shape by [x,y,z] degrees about origin of [x,y,x]. "
        r_rads = r_angles * np.pi / np.float(180.0)
        x_rot = np.mat([ [                1.0,                0.0,                0.0]
                       , [                0.0,  np.cos(r_rads[0]), -np.sin(r_rads[0])]
                       , [                0.0,  np.sin(r_rads[0]),  np.cos(r_rads[0])]
                       ])
        y_rot = np.mat([ [  np.cos(r_rads[1]),                 0.0,  np.sin(r_rads[1])]
                       , [                0.0,                 1.0,                0.0]
                       , [ -np.sin(r_rads[1]),                 0.0,  np.cos(r_rads[1])]
                       ])
        z_rot = np.mat([ [  np.cos(r_rads[2]), -np.sin(r_rads[2]),                 0.0]
                       , [  np.sin(r_rads[2]),  np.cos(r_rads[2]),                 0.0]
                       , [                0.0,                0.0,                 1.0]
                       ])
        rot_matrix = x_rot * y_rot * z_rot
        new_points = np.array([ (origin + (point - origin) * rot_matrix).A1 
                                for point in self.points() 
                              ])
        self.set_points(new_points)
        return self

class Obscuring():
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def contains_point(self, point):
        " Test if point is within the shape. "
        pass


class Line(Rotateable):
    def __init__(self, A, B):
        self.A = np.array(A)
        self.B = np.array(B)

        self.center = np.array(( np.mean(map(lambda x: x[0], [A,B]))
                               , np.mean(map(lambda x: x[1], [A,B]))
                               , np.mean(map(lambda x: x[2], [A,B]))
                               ))

    def set_points(self, points):
        self.__init__(points[0], points[1])
        return self

    def shift(self, offset):
        self.__init__( self.A + offset
                     , self.B + offset
                     )
        return self

    def points(self):
        return [self.A, self.B]

    def fill(self, N=4):
        vec = self.B - self.A
        return [self.A + vec * ii for ii in np.linspace(0,1,N)]


class Tri(Rotateable, Obscuring):
    def __init__(self, A, B, C):
        self.A = np.array(A)
        self.B = np.array(B)
        self.C = np.array(C)

        self.AB = self.B - self.A
        self.AC = self.C - self.A

        self.center = np.array(( np.mean(map(lambda x: x[0], [A,B,C]))
                               , np.mean(map(lambda x: x[1], [A,B,C]))
                               , np.mean(map(lambda x: x[2], [A,B,C]))
                               ))

    def set_points(self, points):
        self.__init__(points[0], points[1], points[2])
        return self

    def shift(self, offset):
        self.__init__( self.A + offset
                     , self.B + offset
                     , self.C + offset
                     )
        return self

    def points(self):
        return [self.A, self.B, self.C]

    def point_in_plane(self, u, v):
        return self.A + u * self.AC + v * self.AB

    def point_in_triangle(self, p):
        v0 = self.AC
        v1 = self.AB
        v2 = np.array(p) - self.A

        dot00 = np.dot(v0, v0)
        dot01 = np.dot(v0, v1)
        dot02 = np.dot(v0, v2)
        dot11 = np.dot(v1, v1)
        dot12 = np.dot(v1, v2)

        denom = (dot00 * dot11 - dot01 * dot01)
        u = (dot11 * dot02 - dot01 * dot12) / denom
        v = (dot00 * dot12 - dot01 * dot02) / denom

        return (u >= 0) and (v >= 0) and (u + v <= 1)

    def contains_point(self, p): 
        return self.point_in_triangle(p)

    def fill(self, N=4):
        us = np.linspace(0,1,N)
        vs = np.linspace(0,1,N)

        new_points = [ self.point_in_plane(u,v) 
                       for (u,v) in product(us, vs) 
                       if u + v <= 1.0 
                     ]

        return new_points

    def __str__(self):
        return "{},{},{}".format(self.A, self.B, self.C)


class Tetrahedron(Rotateable, Obscuring):
    def __init__(self, A, B, C, D):
        self.A = np.array(A)
        self.B = np.array(B)
        self.C = np.array(C)
        self.D = np.array(D)

        self.center = np.array(( np.mean(map(lambda x: x[0], [A,B,C,D]))
                               , np.mean(map(lambda x: x[1], [A,B,C,D]))
                               , np.mean(map(lambda x: x[2], [A,B,C,D]))
                               ))

        self.AB = self.B - self.A
        self.AC = self.C - self.A
        self.AD = self.D - self.A

    def set_points(self, points):
        self.__init__(points[0], points[1], points[2], points[3])
        return self

    def shift(self, offset):
        self.__init__( self.A + offset
                     , self.B + offset
                     , self.C + offset
                     , self.D + offset
                     )
        return self

    def points(self):
        return [self.A, self.B, self.C, self.D]


    def contains_point(self, p): 
        ss0 = sameside3d(p, self.A, self.B, self.C, self.D)
        ss1 = sameside3d(p, self.B, self.A, self.C, self.D)
        ss2 = sameside3d(p, self.C, self.B, self.B, self.D)
        ss3 = sameside3d(p, self.D, self.A, self.B, self.C)
        return all([ss0, ss1, ss2, ss3])

    def point_in_space(self, u, v, w):
        return self.A + u * self.AC + v * self.AB + w * self.AD


    def fill(self, N=4):
        us = np.linspace(0,1,N)
        vs = np.linspace(0,1,N)
        ws = np.linspace(0,1,N)

        new_points = [ self.point_in_space(u,v,w) 
                       for (u,v,w) in product(us, vs, ws) 
                       if u + v + w <= 1.0 
                     ]

        return new_points


    def shell_vis(self, eye, N=4, bottom=True):
        result = []
        seeA =  sameside3d(eye, self.A, self.B, self.C, self.D)
        seeB =  sameside3d(eye, self.B, self.C, self.D, self.A)
        seeC =  sameside3d(eye, self.C, self.D, self.A, self.B)
        seeD =  sameside3d(eye, self.D, self.A, self.B, self.C)

        if not seeD : result.extend(Tri(self.A,self.B,self.C).fill(N))
        if not seeB : result.extend(Tri(self.A,self.C,self.D).fill(N))
        if not seeC : result.extend(Tri(self.A,self.B,self.D).fill(N))
        if not seeA and bottom : result.extend(Tri(self.B,self.C,self.D).fill(N))
        return result


    def shell(self, N=4, bottom=True):
        result = []
        result.extend(Tri(self.A,self.B,self.C).fill(N))
        result.extend(Tri(self.A,self.B,self.D).fill(N))
        result.extend(Tri(self.A,self.C,self.D).fill(N))
        if bottom:
            result.extend(Tri(self.B,self.C,self.D).fill(N))
        return result


    def __str__(self):
        return "{},{},{},{}".format(self.A, self.B, self.C, self.D)


class Sphere(Rotateable, Obscuring):
    def __init__(self, center, radius, N=12):
        self.center  = np.array(center)
        self.radius  = np.array(radius)
        self._points = self.points(N, initilize=True)

    def points(self, N=5, initilize=False): 
        " Return constructible points in a list. "
        if not initilize: return self._points

        lats = np.array(np.linspace(-np.pi/2, np.pi/2, N))
        lons = np.array(np.linspace(  -np.pi,   np.pi, N))

        result = np.ndarray(shape=(len(lats) * len(lons), 3), dtype=np.float32)

        for (ii, (lat, lon)) in enumerate(product(lats, lons)):
            result[ii] = np.array(( self.radius * np.cos(lat) * np.cos(lon)
                                  , self.radius * np.cos(lat) * np.sin(lon)
                                  , self.radius * np.sin(lat)
                                  )) + self.center
        return result

    def set_points(self, points):
        " Construct shape from list of points. "
        self._points = points
        return self

    def contains_point(self, point):
        " Test if point is within the shape. "
        return np.linalg.norm(point-center) < self.radius

    def fill(self, N=5):
        return self.points(N)

    def shell_vis(self, eye, N=5):
        np_eye = np.array(eye)
        dist = np.linalg.norm(self.center - np_eye)
        return [ point for point in self.points(N) if np.linalg.norm(point - np_eye) < dist ]

    def shell(self, N=5):
        return self.points(N)

    def clear_points(self):
        self._points = []
        return self

    def set_lat_lon_points(self, lat_lon_points):
        self._points = np.ndarray(shape=(len(lat_lon_points), 3), dtype=np.float32)
        for (ii, point) in enumerate(lat_lon_points):
            self._points[ii] = np.array(( self.radius * np.cos(point.lat * np.pi/180.0) * np.cos(point.lon * np.pi/180.0)
                                        , self.radius * np.cos(point.lat * np.pi/180.0) * np.sin(point.lon * np.pi/180.0)
                                        , self.radius * np.sin(point.lat * np.pi/180.0)
                                        )) + self.center
        return self

    def __str__(self):
        return "{} {} {}".format(self.center, self.radius, self.points())


class Cube(Rotateable, Obscuring):
    def __init__(self, center, half_width):
        self.center = np.array(center)
        self.middle = Tetrahedron( (-half_width,-half_width,-half_width)
                                 , ( half_width, half_width,-half_width)
                                 , (-half_width, half_width, half_width)
                                 , ( half_width,-half_width, half_width)
                                 )

        self.A = Tetrahedron( -self.middle.A, self.middle.B, self.middle.C, self.middle.D )
        self.B = Tetrahedron( -self.middle.B, self.middle.A, self.middle.C, self.middle.D )
        self.C = Tetrahedron( -self.middle.C, self.middle.A, self.middle.B, self.middle.D )
        self.D = Tetrahedron( -self.middle.D, self.middle.A, self.middle.B, self.middle.C )

        self.middle.shift(center)
        self.A.shift(center)
        self.B.shift(center)
        self.C.shift(center)
        self.D.shift(center)

    def set_points(self, points):
        self.middle.set_points((points[0], points[1], points[2], points[3]))
        self.A.set_points((points[4], points[1], points[2], points[3]))
        self.B.set_points((points[5], points[0], points[2], points[3]))
        self.C.set_points((points[6], points[0], points[1], points[3]))
        self.D.set_points((points[7], points[0], points[1], points[2]))
        return self

    def points(self):
        result = self.middle.points()
        result.append(self.A.A)
        result.append(self.B.A)
        result.append(self.C.A)
        result.append(self.D.A)
        return result

    def contains_point(self, p): 
        return any(( self.middle.contains_point(p)
                   , self.A.contains_point(p)
                   , self.B.contains_point(p)
                   , self.C.contains_point(p)
                   , self.D.contains_point(p)
                   ))

    def fill(self, N=4):
        result = self.middle.fill(N)
        result.extend(self.A.fill(N))
        result.extend(self.B.fill(N))
        result.extend(self.C.fill(N))
        result.extend(self.D.fill(N))
        return result

    def shell_vis(self, eye, N=4):
        result = []
        result.extend(self.A.shell_vis(eye, N, bottom=False))
        result.extend(self.B.shell_vis(eye, N, bottom=False))
        result.extend(self.C.shell_vis(eye, N, bottom=False))
        result.extend(self.D.shell_vis(eye, N, bottom=False))
        return result

    def shell(self, N=4):
        result = []
        result.extend(self.A.shell(N, bottom=False))
        result.extend(self.B.shell(N, bottom=False))
        result.extend(self.C.shell(N, bottom=False))
        result.extend(self.D.shell(N, bottom=False))
        return result

    def __str__(self):
        return "{}".format(self.points())


if __name__=='__main__':

    S = Sphere((0,0,0), 5)

    print S

    # C = Cube(np.array([0,0,0]), 5)

    # p0 = np.array([0.0, 0.0, 0.0])
    # p1 = np.array([7.2, 0.0, 0.0])
    # p2 = np.array([  0, 7.2, 0.0])
    # p3 = np.array([  0,   0, 7.2])
    # p4 = np.array([0.7, 1.0, 0.0])


    # print C.contains_point(p0)
    # print C.contains_point(p1)
    # print C.contains_point(p2)
    # print C.contains_point(p3)
    # print C.contains_point(p4)

    # print C.shell_vis(np.array([0,0,-30]))

File: faster/src/shapes.cpp
#include <shapes.h>

Vec3 add     (Vec3 A, Vec3 B)     { return {A[0]+B[0],A[1]+B[1],A[2]+B[2]}; };
Vec3 sub     (Vec3 A, Vec3 B)     { return {A[0]-B[0],A[1]-B[1],A[2]-B[2]}; };

Vec3 scale   (double a,   Vec3 B) { return {a*B[0],a*B[1],a*B[2]}; };
Vec3 scale   (  Vec3 B, double a) { return {a*B[0],a*B[1],a*B[2]}; };

double dot   (Vec3 A, Vec3 B)     { return (A[0]*B[0])+(A[1]*B[1])+(A[2]*B[2]); };
Vec3   cross (Vec3 A, Vec3 B)     { 
    return { (A[1]*B[2])-(A[2]*B[1]) 
           , (A[2]*B[0])-(A[0]*B[2]) 
           , (A[0]*B[1])-(A[1]*B[0]) 
           };
};

bool sameside3d( Vec3 p    // Perspective Point
               , Vec3 ref  // Reference Point
               , Vec3 A, Vec3 B, Vec3 C  // Shape Plane points
               )
{
    Vec3   cpAB_BC  = cross(sub(A,B), sub(A,C));
    double sameABCr = dot(sub(ref,A), cpAB_BC);
    double sameABCp = dot(sub(  p,A), cpAB_BC);

    if ( sameABCr < 0 && sameABCp < 0 ) return true;
    if ( sameABCr > 0 && sameABCp > 0 ) return true;

    return false;
}

Vec3 perspective_division( Vec3   & eye
                         , Vec3     point
                         , double   plane_dist
                         )
{
    point[2] = point[2] - eye[2];
    for(int ii=0; ii<2; ++ii)
      point[ii] = (point[ii] - eye[ii]) * plane_dist/point[2];
    return point;
};

deque<Vec3> Shape::perspectiveDivision(Vec3 & eye, double plane_dist)
{
    deque<Vec3> result;
    for(Vec3 point : this->getPoints(eye))
        result.push_back(perspective_division(eye, point, plane_dist));

    return result;
}


void Shape::setRotation( Vec3   RotationVector
                       , double RotationAngle
                       )
{
    double sqSum = sqrt(RotationVector[0]+RotationVector[1]+RotationVector[2]);
    double halfAngle = RotationAngle/2.0;
    double sinHfAngl = sin(halfAngle);
    q_RotationVector = Quaternion<double>( cos(halfAngle)
                                         , (RotationVector[0]/sqSum)*sinHfAngl
                                         , (RotationVector[1]/sqSum)*sinHfAngl
                                         , (RotationVector[2]/sqSum)*sinHfAngl
                                         );
};


// Rotate the shape by RotationVector and RotationAngle
void Shape::rotate()
{
    double new_point[3];
    int ii;
    for(auto &point : points)
    {
        for (ii=0; ii<3; ++ii) new_point[ii] = point[ii] - origin[ii];

        q_RotationVector.QuatRotation(new_point);

        for (ii=0; ii<3; ++ii) point[ii] = new_point[ii] + origin[ii];
    }
};

void Shape::shift(Vec3 offset) 
{
    for(auto &point : points)
    {
        point[0] += offset[0];
        point[1] += offset[1];
        point[2] += offset[2];
    }
    origin[0] += offset[0];
    origin[1] += offset[1];
    origin[2] += offset[2];
};

Line::Line(Vec3 A, Vec3 B, int N)
{
    double deltaX = (B[0] - A[0])/N;
    double deltaY = (B[1] - A[1])/N;
    double deltaZ = (B[2] - A[2])/N;
    for(int index=0; index<=N; ++index)
    {
        Vec3 point;
        point.push_back(A[0] + (index * deltaX));
        point.push_back(A[1] + (index * deltaY));
        point.push_back(A[2] + (index * deltaZ));
        this->points.push_back(point);
    }
    this->origin = A;
};


Tri::Tri(Vec3 & A, Vec3 & B, Vec3 & C, int N)
{
    this->N = N;

    points.push_back(A);
    points.push_back(B);
    points.push_back(C);

    setOrigin({ (points[0][0]+points[1][0]+points[2][0])/3.0
              , (points[0][1]+points[1][1]+points[2][1])/3.0
              , (points[0][2]+points[1][2]+points[2][2])/3.0
              });
};

deque<Vec3> Tri::fill()
{
    Vec3 AB = sub(points[1], points[0]);
    Vec3 AC = sub(points[2], points[0]);

    deque<Vec3> result;
    for(double u=0.0; u<=1.0; u+=0.99/N)
    {
        for(double v=0.0; v<=1.0; v+=1.0/N)
        {
            if ( (u+v) >= 1.0 ) continue;
            result.push_back(add(points[0],add(scale(u,AC),scale(v,AB))));
        }
    }
    return result;
};


Tetrahedron::Tetrahedron()
{
    N = 0;
    points.push_back({0,0,0});
    points.push_back({0,0,0});
    points.push_back({0,0,0});
    points.push_back({0,0,0});
    setOrigin({0,0,0});
};

Tetrahedron::Tetrahedron(const Tetrahedron &original)
{
    this->setA(original.getA());
    this->setB(original.getB());
    this->setC(original.getC());
    this->setD(original.getD());
    this->setN(original.getN());

    setOrigin({ (points[0][0]+points[1][0]+points[2][0]+points[3][0])/4.0
              , (points[0][1]+points[1][1]+points[2][1]+points[3][1])/4.0
              , (points[0][2]+points[1][2]+points[2][2]+points[3][2])/4.0
              });
};

Tetrahedron::Tetrahedron(Vec3 A, Vec3 B, Vec3 C, Vec3 D, int N)
{
    this->N = N;
    points.push_back(A);
    points.push_back(B);
    points.push_back(C);
    points.push_back(D);

    setOrigin({ (points[0][0]+points[1][0]+points[2][0]+points[3][0])/4.0
              , (points[0][1]+points[1][1]+points[2][1]+points[3][1])/4.0
              , (points[0][2]+points[1][2]+points[2][2]+points[3][2])/4.0
              });
};


deque<Vec3> Tetrahedron::shell_vis(Vec3 & eye, bool bottom)
{
    deque<Vec3> result;
    bool seeA =  sameside3d(eye, points[0], points[1], points[2], points[3]);
    bool seeB =  sameside3d(eye, points[1], points[2], points[3], points[0]);
    bool seeC =  sameside3d(eye, points[2], points[3], points[0], points[1]);
    bool seeD =  sameside3d(eye, points[3], points[0], points[1], points[2]);

    if (!seeD)           for(auto const &point :Tri(points[0],points[1],points[2], N).fill()) { result.push_back(point); }
    if (!seeB)           for(auto const &point :Tri(points[0],points[2],points[3], N).fill()) { result.push_back(point); }
    if (!seeC)           for(auto const &point :Tri(points[0],points[1],points[3], N).fill()) { result.push_back(point); }
    if (!seeA && bottom) for(auto const &point :Tri(points[1],points[2],points[3], N).fill()) { result.push_back(point); }

    return result;
};

Cube::Cube(const Cube& original)
{
    this->A      = original.A;
    this->B      = original.B;
    this->C      = original.C;
    this->D      = original.D;
    this->middle = original.middle;
};

Cube::Cube(Vec3 center, double side_length, int N)
{
    double half_width = side_length/2;
    this->setOrigin(center);

    middle = Tetrahedron( {-half_width,-half_width,-half_width}
                        , { half_width, half_width,-half_width}
                        , {-half_width, half_width, half_width}
                        , { half_width,-half_width, half_width}
                        , N
                        );

    A = Tetrahedron( scale(-1, middle.getA()), middle.getB(), middle.getC(), middle.getD(), N);
    B = Tetrahedron( scale(-1, middle.getB()), middle.getA(), middle.getC(), middle.getD(), N);
    C = Tetrahedron( scale(-1, middle.getC()), middle.getA(), middle.getB(), middle.getD(), N);
    D = Tetrahedron( scale(-1, middle.getD()), middle.getA(), middle.getB(), middle.getC(), N);

    middle.shift(center);
    A.shift(center);
    B.shift(center);
    C.shift(center);
    D.shift(center);

    A.setOrigin(center);
    B.setOrigin(center);
    C.setOrigin(center);
    D.setOrigin(center);
};


deque<Vec3> Cube::getPoints(Vec3 & eye) 
{ 
    deque<Vec3> result;
    for(auto const &point : A.shell_vis(eye, false)) result.push_back(point);
    for(auto const &point : B.shell_vis(eye, false)) result.push_back(point);
    for(auto const &point : C.shell_vis(eye, false)) result.push_back(point);
    for(auto const &point : D.shell_vis(eye, false)) result.push_back(point);
    return result;
};

void Cube::rotate()
{
    middle.rotate();
    A.rotate();
    B.rotate();
    C.rotate();
    D.rotate();
};

void Cube::setRotation(Vec3 RotationVector, double RotationAngle)
{
    middle.setRotation(RotationVector, RotationAngle);
    A.setRotation(RotationVector, RotationAngle);
    B.setRotation(RotationVector, RotationAngle);
    C.setRotation(RotationVector, RotationAngle);
    D.setRotation(RotationVector, RotationAngle);
};

